src/mc-fle2-rfds.c

```diff
@@ -0,0 +1,615 @@
+      TRY (BSON_APPEND_UTF8 (&operator, "0", spec->field));
+      TRY (_mongocrypt_buffer_append (&p1, &operator, "1", 1));
+      TRY (bson_append_array_end (&elem, &operator));
+      TRY (bson_append_document_end (&and, &elem));
+
+      if (spec->nOps == 2) {
+         TRY (BSON_APPEND_DOCUMENT_BEGIN (&and, "1", &elem));
+         TRY (BSON_APPEND_ARRAY_BEGIN (
+            &elem, mc_FLE2RangeOperator_to_string (spec->secondOp), &operator));
+         TRY (BSON_APPEND_UTF8 (&operator, "0", spec->field));
+         TRY (_mongocrypt_buffer_append (&p2, &operator, "1", 1));
+         TRY (bson_append_array_end (&elem, &operator));
+         TRY (bson_append_document_end (&and, &elem));
+      }
+
+      TRY (bson_append_array_end (out, &and));
+   } else {
+      /*
+      Create a Match Expression document like:
+      {
+        "$and" : [
+            {"age" : { "$gt": "<placeholder1>" }},
+            {"age" : { "$lt": "<placeholder2>" }}
+         ]
+      }
+      */
+      bson_t and;
+      TRY (BSON_APPEND_ARRAY_BEGIN (out, "$and", &and));
+      bson_t elem;
+      TRY (BSON_APPEND_DOCUMENT_BEGIN (&and, "0", &elem));
+      bson_t operator;
+      TRY (BSON_APPEND_DOCUMENT_BEGIN (&elem, spec->field, &operator));
+      const char *op_str = mc_FLE2RangeOperator_to_string (spec->firstOp);
+      TRY (_mongocrypt_buffer_append (&p1, &operator, op_str, -1));
+      TRY (bson_append_document_end (&elem, &operator));
+      TRY (bson_append_document_end (&and, &elem));
+
+      if (spec->nOps == 2) {
+         TRY (BSON_APPEND_DOCUMENT_BEGIN (&and, "1", &elem));
+         TRY (BSON_APPEND_DOCUMENT_BEGIN (&elem, spec->field, &operator));
+         const char *op_str = mc_FLE2RangeOperator_to_string (spec->secondOp);
+         TRY (_mongocrypt_buffer_append (&p2, &operator, op_str, -1));
+         TRY (bson_append_document_end (&elem, &operator));
+         TRY (bson_append_document_end (&and, &elem));
+      }
+
+      TRY (bson_append_array_end (out, &and));
+   }
+
+#undef TRY
+   ok = true;
+fail:
+   _mongocrypt_buffer_cleanup (&p2);
+   _mongocrypt_buffer_cleanup (&p1);
+   bson_destroy (&infDoc);
+   return ok;
+}
+
+static mlib_once_flag payloadId_init_flag = MLIB_ONCE_INITIALIZER;
+static mongocrypt_mutex_t payloadId_mutex;
+static int32_t payloadId = 0;
+
+static void
+payloadId_init_mutex (void)
+{
+   _mongocrypt_mutex_init (&payloadId_mutex);
+}
+
+void
+mc_reset_payloadId_for_testing (void)
+{
+   mlib_call_once (&payloadId_init_flag, payloadId_init_mutex);
+   MONGOCRYPT_WITH_MUTEX (payloadId_mutex)
+   {
+      payloadId = 0;
+   }
+}
+
+// mc_getNextPayloadId is thread safe.
+int32_t
+mc_getNextPayloadId (void)
+{
+   mlib_call_once (&payloadId_init_flag, payloadId_init_mutex);
+   int32_t ret;
+   MONGOCRYPT_WITH_MUTEX (payloadId_mutex)
+   {
+      if (payloadId >= INT32_MAX) {
+         payloadId = 0;
+      }
+      ret = payloadId;
+      payloadId++;
```

Comment:
```suggestion
      ret = payloadId;
      payloadId = payloadId < INT32_MAX ? payloadId + 1 : 0;
```
`payloadId > INT32_MAX` is an impossible condition. The value `INT32_MAX` is skipped as-written. To include `INT32_MAX` as a possible payloadId, recommend rewriting as suggested.